src/garage/envs/environment.py

```diff
@@ -0,0 +1,188 @@
+                  See render() for a list of modes.
+
+        """
+
+    @abc.abstractmethod
+    def reset(self):
+        """Resets the environment and returns the first `TimeStep` of sequence.
+
+        Returns:
+            garage.TimeStep: The first TimeStep.
+                Note that this `TimeStep` will have attributes `observation`,
+                `action`, `reward` as None as they are undefined.
+
+        """
+
+    @abc.abstractmethod
+    def step(self, action):
+        """Steps the environment using to the action and returns a `TimeStep`.
+
+        If the environment returned the last time step of a sequence (either
+        because of terminal or time out) at the previous step, this call to
+        `step()` will start a new sequence and `action` will be ignored.
+
+        This method will also start a new sequence if called after the
+        environment has been constructed and `reset()` has not been called.
+        Again, in this case `action` will be ignored.
+
+        If possible, update the visualization display as well.
+
+        Args:
+            action (np.ndarray): A NumPy array, or a nested dict, list or tuple
+                of arrays corresponding to `action_spec()`.
+
+        Returns:
+            garage.TimeStep: The TimeStep after the action.
+
+        """
+
+    @abc.abstractmethod
+    def render(self, mode):
+        """Renders the environment.
+
+        The set of supported modes varies per environment. By convention,
+        if mode is:
+
+        * rgb_array: Return an numpy.ndarray with shape (x, y, 3),
+          representing RGB values for an x-by-y pixel image, suitable
+          for turning into a video.
+        * ansi: Return a string (str) or StringIO.StringIO containing a
+          terminal-style text representation. The text can include newlines
+          and ANSI escape sequences (e.g. for colors).
+
+        Make sure that your class's `render_modes` includes the list of
+        supported modes.
+
+        For example:
+
+            class MyEnv(Env):
+                def render_modes(self):
+                    return ['rgb_array', 'ansi']
+
+                def render(self, mode):
+                    if mode == 'rgb_array':
+                        return np.array(...) # return RGB frame for video
+                    elif mode == 'ansi':
+                        ... # return text output
+                    else:
+                        super(MyEnv, self).render(mode=mode) # just raise an
+                        exception
+
+        Args:
+            mode (str): the mode to render with. The string must be present in
+                `self.render_modes`.
+
+        """
+
+    @abc.abstractmethod
+    def visualize(self):
+        """Creates a visualization of the environment.
+
+        This function should be called only once after reset() to set up the
+        visualization display. Update of the visualization should be done in
```

Comment:
i think you should make very explicit here that calling "close()" should deallocate any resources and close any windows created by calling this. 

you may consider providing a way of closing the visualization without closing the environment. 

@krzentner @ahtsan WDYT?

Reply:
I think "closing the visualization without closing the environment" is definitely a common use case. However, I am not sure if such functionality should be placed in the main API. 

For instance if an environment opens many visualization windows, then something like `close_visualization()` will need to determine which window to close. It may also need extra parameters.

These details can be left to subclasses implementations. In my opinion the main API should do less and simply say that close() will close _all_ visualizations created by `visualize()`.

Reply:
that's fine with me. it can always be revised.

in the case of windows, a user can just click [x] :)